Specification.EntityFrameworkCore/src/Ardalis.Specification.EntityFrameworkCore/ContextFactoryRepositoryBaseOfT.cs

```diff
@@ -0,0 +1,227 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.EntityFrameworkCore;
+
+namespace Ardalis.Specification.EntityFrameworkCore
+{
+  public abstract class ContextFactoryRepositoryBaseOfT<TEntity, TContext> : IRepositoryBase<TEntity> 
+    where TEntity : class
+    where TContext : DbContext
+  {
+    private IDbContextFactory<TContext> dbContextFactory;
+    private ISpecificationEvaluator specificationEvaluator;
+
+    public ContextFactoryRepositoryBaseOfT(IDbContextFactory<TContext> dbContextFactory)
+      : this(dbContextFactory, SpecificationEvaluator.Default)
+    {
+    }
+
+    public ContextFactoryRepositoryBaseOfT(IDbContextFactory<TContext> dbContextFactory,
+      ISpecificationEvaluator specificationEvaluator)
+    {
+      this.dbContextFactory = dbContextFactory;
+      this.specificationEvaluator = specificationEvaluator;
+    }
+
+    /// <inheritdoc/>
+    public async Task<TEntity?> GetByIdAsync<TId>(TId id, CancellationToken cancellationToken = default) where TId : notnull
+    {
+      await using var dbContext = this.dbContextFactory.CreateDbContext();
+      return await dbContext.Set<TEntity>().FindAsync(new object[] { id }, cancellationToken: cancellationToken);
+    }
+
+    /// <inheritdoc/>
+    public async Task<TEntity?> GetBySpecAsync(ISpecification<TEntity> specification, CancellationToken cancellationToken = default)
+    {
+      await using var dbContext = this.dbContextFactory.CreateDbContext();
+      return await ApplySpecification(specification, dbContext).FirstOrDefaultAsync(cancellationToken);
+    }
+
+    /// <inheritdoc/>
+    public async Task<TResult?> GetBySpecAsync<TResult>(ISpecification<TEntity, TResult> specification, CancellationToken cancellationToken = default)
+    {
+      await using var dbContext = this.dbContextFactory.CreateDbContext();
+      return await ApplySpecification(specification, dbContext).FirstOrDefaultAsync(cancellationToken);
+    }
+
+    /// <inheritdoc/>
+    public async Task<TEntity?> FirstOrDefaultAsync(ISpecification<TEntity> specification, CancellationToken cancellationToken = default)
+    {
+      await using var dbContext = this.dbContextFactory.CreateDbContext();
+      return await ApplySpecification(specification, dbContext).FirstOrDefaultAsync(cancellationToken);
+    }
+
+    /// <inheritdoc/>
+    public async Task<TResult?> FirstOrDefaultAsync<TResult>(ISpecification<TEntity, TResult> specification, CancellationToken cancellationToken = default)
+    {
+      await using var dbContext = this.dbContextFactory.CreateDbContext();
```

Comment:
Apologies Steve, I didn't give that fact much attention in my initial request.

Yes, I would expect such an error with this approach. I'll add a virtual protected method which invokes the `TrackGraph` method inside the `using` to solve this. The original intention was to allow for maximum flexibility but on reflection a default approach to managing the change tracker is definitely appropriate.

I just need to work through getting the error we're expecting to present itself in tests to ensure it's being handled appropriately.

I'll post another update here when the above is completed.

Reply:
>I'll add a virtual protected method which invokes the TrackGraph method inside the using to solve this.

Was this done? If so, where? Either way, can you explain this for me please?

Thanks,

Matt

Reply:
@mwasson74, this wasn't completed in the end, mainly because the only solutions I could come up with added too much opinion into the library.

Essentially, due to the prolonged lifecycle of Blazor (and WPF/UWP/MAUI etc) apps, the new ContextFactoryRepositoryBaseOfT class instantiates a new instance of the DBContext every time a method is invoked. This means that all of the Entity Framework Change Tracking goodness is lost.

From a DDD perspective, this actually makes sense, whether or not an Entity has changed is a subject for the Entity to manage itself and not be reliant on a 3rd party to deduce.

What this means is that you need to implement change tracking manually within your solution. I think the only method you'll have to overload is the UpdateAsync method since that's the only one which really needs to know what's changed.

Sorry this is so vague, it's been a while since I've looked at this. Reply back here if you need more info. I'll try and carve out some time to write a proper doc for this as well.